File: 2022/03/05/longestPalindrome.ts
/**
 * https://leetcode.com/problems/longest-palindrome/
 * @author Andrew Jarombek
 * @since 3/5/2022
 */

function longestPalindrome(s: string): number {
    const chars = new Map<string, number>();

    for (let c of s) {
        chars.set(c, (chars.get(c) ?? 0) + 1);
    }

    let total = 0;
    let odd = 0;
    let oddChar = '';

    for (let [key, val] of chars.entries()) {
        if (val % 2 == 0) {
            total += val;
        } else if (val > odd) {
            total += Math.max(0, odd - 1);
            odd = val;
            oddChar = key;
        } else {
            total += val - 1;
        }
    }

    return total + odd;
}

File: 2023/06/18/go/longest_palindrome.go
/**
 * https://leetcode.com/problems/longest-palindrome/
 * Author: Andrew Jarombek
 * Date: 6/18/2023
 */

package _go

func longestPalindrome(s string) int {
	var chars = make(map[rune]int)

	for _, c := range s {
		chars[c]++
	}

	result := 0
	odd := false

	for _, value := range chars {
		if value%2 == 1 {
			odd = true
			result += value - 1
		} else {
			result += value
		}
	}

	if odd {
		result++
	}

	return result
}
